src/Umbraco.Core/Models/PublishedContent/PublishedContentTypeFactory.cs

```diff
@@ -73,24 +66,31 @@ public PublishedDataType GetDataType(int id)
     }
 
     /// <inheritdoc />
-    public void NotifyDataTypeChanges(int[] ids)
+    public void NotifyDataTypeChanges(params int[] ids)
     {
+        if (_publishedDataTypes is null)
+        {
+            // Not initialized yet, so skip and avoid lock
+            return;
+        }
+
         lock (_publishedDataTypesLocker)
         {
-            if (_publishedDataTypes == null)
+            if (ids.Length == 0)
             {
-                IEnumerable<IDataType> dataTypes = _dataTypeService.GetAll();
-                _publishedDataTypes = dataTypes.ToDictionary(x => x.Id, CreatePublishedDataType);
+                // Clear cache (and let it lazy initialize again later)
+                _publishedDataTypes = null;
```

Comment:
It's better for performance to clear the cached `PublishedDataType`s, so this method can be called multiple times without fetching all data types (and in subsequent calls without taking the lock).